### DISCUSSION TITLE
Help needed: How should I interpret the PS processing output in SAR geometry?

### DISCUSSION THREAD

**Discussion by Shima-shoki** [isce-framework/dolphin#506] [Q&A] [answered]
Hi, I was able to run the dolphin by using the coregistered SLC files by using ISCE2. I believe that my files are in SAR geometry and there is no geometric transforms in *.vrt files. I chose Jakarta area as an example of InSAR processing. Some intermediate files such as the amp_mean.tif (Fig. 1) and ps_pixels (Fig. 2) looked okay as shown below. **The question is, the final outcome of the PS processing, such as the velocity, has a strange unit.** Based on a previous discussion #503 , the unit of the dolphin output is in meters. However, the output shows the unit in -20 ~ 0 based on the plot, which is unlikely in a usual subsidence events. The velocity output also has a range from -11 ~ 10. 

My guess is that if a SAR geometry data is supplied to the software, the output is calculated in the unit of radians, but I am still wondering how to correctly interpret the output in this case. I also really appreciate to know whether dolphin can make a geocoded output as a final product. Any kind help is very much appreciated!

The log & config file: [dolphin_log-config.zip](https://github.com/user-attachments/files/17975955/dolphin_log-config.zip)

![image](https://github.com/user-attachments/assets/4b8274ba-f498-4df3-be2a-a4681986d1aa)
Figure 1. "amp_mean.tif"

![image](https://github.com/user-attachments/assets/f143433f-e2f9-4656-8413-25c3e354813e)
Figure 2. "velocity.tif"

![image](https://github.com/user-attachments/assets/42ca4386-fbeb-48dd-92b0-37b7321ac40a)
Figure 3. A visualization of displacement following the "basic walkthrough" example. The "masked" parameter was changed from "True" to "False", since the visualization was not shown if it is set to "True".

**Comment by scottstanie** [comment]
Hi I haven't checked the logs, but right now there's no attempt to automatically recognize the sensor of input data; therefore the output would be in radians if you don't add "wavelength" to the config.
(OPERA sentinel 1 does have a switch now, since it's been the main test cases. Considering adding more later, but unsure now)

And if the "masked" argument for loading shows nothing, I believe that means it is a nodata value... do you mean the time series is masked? Or it's masked based on the connected component labels?

**Reply by Shima-shoki** [reply]
@scottstanie 
Thank you very much for the quick reply!

>there's no attempt to automatically recognize the sensor of input data; therefore the output would be in radians if you don't add "wavelength" to the config.  

Well noted about the output of the PS processing. I rechecked the documents, and there is a ["wavelength" option in the code reference as per the advice](https://dolphin-insar.readthedocs.io/en/latest/reference/dolphin/workflows/config/). However, I could not find it when I called "dolphin config -h". Is it an intentional behavior of the software ...? (sorry if I missed it in the help!)

>And if the "masked" argument for loading shows nothing, I believe that means it is a nodata value... do you mean the time series is masked? Or it's masked based on the connected component labels?  

Sorry my question was vague where the "masked" option was changed. This is the code I used to visualize the final output as per the "basic walkthrough" instruction. I believe that there should be some kind of output values in the file, but my understanding of the Dolphin is yet to be quite limited...

```python
fig, axes = plt.subplots(ncols=2, figsize=(20, 6))
ax = axes[0]

last_displacement = load_gdal(file_list[-1], masked=True)#The visualization in axes[0] is empty (axes[1] is showing a time-series)

axim = ax.imshow(
    last_displacement - last_displacement[0].mean(), cmap="RdBu", vmax=0, vmin=-20
)
fig.colorbar(axim, label="LOS [radians]")

ax = axes[1]
row, col = np.unravel_index(np.argmin(last_displacement), last_displacement.shape)
dates = [get_dates(f)[1] for f in file_list]
timeseries_values = [
    load_gdal(f, rows=slice(row, row + 1), cols=slice(col, col + 1)) for f in file_list
]

ax.plot(dates, np.array(timeseries_values).squeeze(), ".-")
```
![image](https://github.com/user-attachments/assets/ef54fb2d-9889-41db-8747-7adc5d314e05)


I also tested the visualization of the "velocity.tif" with masked=True option, then the output looks okay.
```python
velocity=load_gdal('timeseries/velocity.tif', masked=True)
plt.imshow(velocity)
```

![image](https://github.com/user-attachments/assets/8b562159-20d8-42d8-bea9-290338f19cf6)

**Reply by scottstanie** [reply]
Ah I didn't add a clear enough comment- that subtracting of the mean of the first row was just an arbitrary way to spatially reference the image. Your image looks like it has nans in the first row, so I think you're subtracting nan

**Reply by Shima-shoki** [reply]
@scottstanie 
**>that subtracting of the mean of the first row was just an arbitrary way to spatially reference the image**
I see! Thank you for your elaboration on the procedure. I now understood how to interpret the outcome of my processing :)

If possible, would you mind me to give a guidance on processing the final output map (i.e., velocity) to generate a georeferenced image in Dolphin?

**Reply by scottstanie** [reply, marked as answer]
There's no dolphin utility now to do this with images in radar coordinates, but I know that the [UNAVCO tutorial introduction to GDAL](https://github.com/parosen/Geo-SInC/blob/9f3e2858bc4717f3e4c0d2b3e4fe3cc6f69ebf5a/EarthScope2024/0.4_Geographic_Information_Science_using_GDAL/03_RasterProjection.ipynb) demonstrate how to use GDAL along with the `lat` and `lon` files to do so (search for `geocodeUsingGdalWarp` at the bottom)

**Reply by Shima-shoki** [reply]
@scottstanie 
>I know that the [UNAVCO tutorial introduction to GDAL](https://github.com/parosen/Geo-SInC/blob/9f3e2858bc4717f3e4c0d2b3e4fe3cc6f69ebf5a/EarthScope2024/0.4_Geographic_Information_Science_using_GDAL/03_RasterProjection.ipynb) demonstrate how to use GDAL along with the lat and lon files to do so (search for geocodeUsingGdalWarp at the bottom)

Thank you for kindly sharing the information about the UNAVCO tutorial! I'll have a look into the material to make geocoded stacked SLC files to be supplied to Dolphin🐬

**Reply by Karthi-Wolf** [reply]
Hi Mr. Shima shoki, Have you ran dolphin with the geocoded SLCs from ISCE2. When I tried in radar coordinates and exported it into mintpy, it shown, 'input data seems to be geocoded, no need to geocode' but ends up in radar coordinates. Now I'm trying to geocode SLCs like you mentioned. 
Which way is better and efficient? Either processing in radar coordinates or GEO coordinates?

**Reply by scottstanie** [reply]
I think you'll need to provide more detail about exactly what commands you ran on what data in order to have people help debug the mintpy ingesting problem. My guess is that you ran some `prep_?` script that was assuming geo-coordinate images, but more details are needed

**Reply by Karthi-Wolf** [reply]
Hi @scottstanie, Thank you so much and sorry for not elaborating clearly.

Following up with more details.
I'm using ISCE -> Dolphin -> Mintpy. 

I made coregistered SLCs usiing (with less files initially)

"**stackSentinel.py -s/..../slc -o /..../orbits -d /.../dem/demLat_N19_N22_Lon_E084_E087.dem.wgs84 -b '19.760 19.917 85.725 85.935' -a /..../orbits -c 5 -C NESD -W slc -m 20160101 -e 0.30**"

I used Dolphin using ISCE2 coregistered Sentinel-1 SLCs (TOPS stack). 

cslc_file_list:

    **/....../coregistered/merged/SLC/20160101/20160101.slc.full.vrt
    /....../coregistered/merged/SLC/20160113/20160113.slc.full.vrt
    /....../coregistered/merged/SLC/20160125/20160125.slc.full.vrt**

**The inputs are strictly in radar coordinates (no geographic transform in the .vrt)**. I even tired like strides **y : 3 x : 9**.

The dolphin output's dimension were **(1391, 2367)**. After Dolphin finishes, I try to load the outputs into MintPy (ISCE processor mode). 

It's how i made my mintpy config file.

```
mintpy.load.processor        = isce
##---------for ISCE only:
mintpy.load.metaFile         = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/reference/IW*.xml
mintpy.load.baselineDir      = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/baselines
##---------interferogram datasets:
mintpy.load.unwFile          = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/unwrapped/*.unw.tif
mintpy.load.corFile          = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/interferograms/*.int.cor.tif
mintpy.load.connCompFile     = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/unwrapped/*.unw.conncomp.tif
##---------geometry datasets:
mintpy.load.demFile          = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/hgt.rdr
mintpy.load.lookupYFile      = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/lat.rdr
mintpy.load.lookupXFile      = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/lon.rdr
mintpy.load.incAngleFile     = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/los.rdr
mintpy.load.azAngleFile      = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/los.rdr
mintpy.load.shadowMaskFile   = /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/shadowMask.rdr
mintpy.load.waterMaskFile    = None 


```
I initiated the mintpy processes using 

'smallbaselineApp.py mintpy.txt'

I also multilooked the geometryRadar.h5 using multilook.py to match the dolphin strides **y : 3 x : 9**
`multilook.py geometryRadar.h5 -a 3 -r 9`

The ifgramStack.h5 & geometryRadar.h5 size does matches exactly.
The mintpy steps tries loading the file and fails by asking 

`FileNotFoundError: [Errno 2] No such file or directory: '/......./mintpy_1/inputs/geometryGeo.h5`

So, I  tried to geocode the geometryradar.h5 into geometryGeo.h5 usiing '**geocode.py** '

Though I made the geometryGeo.h5 i couldn't force the same size as the geometryRadar.h5 or ifgramStack.h5.
for Example my **geometryRadar.h5 & ifgramStack.h5 size is (1392,2367)** but **geometryGeo.h5 has (269,644)**.
after i corrected gemetryGeo.h5 myself, I got error in the invert_network step 


```
******************** **step - invert_network** ********************
Input data seems to be geocoded. Lookup file not needed.

ifgram_inversion.py /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy/inputs/ifgramStack.h5 -t /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy/smallbaselineApp.cfg --update
read input option from template file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy/smallbaselineApp.cfg
use dataset "unwrapPhase" by default
--------------------------------------------------
update mode: ON
1) output files already exist: ['timeseries.h5', 'temporalCoherence.h5', 'numInvIfgram.h5'].
2) output dataset is newer than input dataset: unwrapPhase.
3) all key configuration parameters are the same: ['obsDatasetName', 'numIfgram', 'weightFunc', 'maskDataset', 'maskThreshold', 'minRedundancy', 'minNormVelocity'].
run or skip: skip.
**Input data seems to be geocoded. Lookup file not needed.**

generate_mask.py /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy/temporalCoherence.h5 -m 0.4 -o /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy/maskTempCoh.h5 --base /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy/inputs/geometryGeo.h5 --base-dataset shadowMask --base-value 1
update mode: ON
run or skip: run
input temporalCoherence file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy/temporalCoherence.h5
read /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy/temporalCoherence.h5
create initial mask with the same size as the input file and all = 1
all pixels with nan value = 0
exclude pixels with value < 0.4
Traceback (most recent call last):
  File "/home/karthi/miniconda3/envs/isce2/bin/smallbaselineApp.py", line 10, in <module>
    sys.exit(main())
             ~~~~^^
  File "/home/karthi/miniconda3/envs/isce2/lib/python3.13/site-packages/mintpy/cli/smallbaselineApp.py", line 209, in main
    run_smallbaselineApp(inps)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/karthi/miniconda3/envs/isce2/lib/python3.13/site-packages/mintpy/smallbaselineApp.py", line 1155, in run_smallbaselineApp
    app.run(steps=inps.runSteps)
    ~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/home/karthi/miniconda3/envs/isce2/lib/python3.13/site-packages/mintpy/smallbaselineApp.py", line 923, in run
    self.run_network_inversion(sname)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/karthi/miniconda3/envs/isce2/lib/python3.13/site-packages/mintpy/smallbaselineApp.py", line 387, in run_network_inversion
    self.generate_temporal_coherence_mask()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/karthi/miniconda3/envs/isce2/lib/python3.13/site-packages/mintpy/smallbaselineApp.py", line 425, in generate_temporal_coherence_mask
    mintpy.cli.generate_mask.main(iargs)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/karthi/miniconda3/envs/isce2/lib/python3.13/site-packages/mintpy/cli/generate_mask.py", line 181, in main
    create_mask(inps)
    ~~~~~~~~~~~^^^^^^
  File "/home/karthi/miniconda3/envs/isce2/lib/python3.13/site-packages/mintpy/generate_mask.py", line 196, in create_mask
    create_threshold_mask(inps)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/karthi/miniconda3/envs/isce2/lib/python3.13/site-packages/mintpy/generate_mask.py", line 151, in create_threshold_mask
    mask[base_data == float(inps.base_value)] = 0
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
**IndexError: boolean index did not match indexed array along axis 0; size of axis is 1392 but size of corresponding boolean axis is 269**
```

**Reply by Karthi-Wolf** [reply]
**Following to that, Mintpy 1.6.2 loads the ISCE + dolphin data somehow and throws** 
"    raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), geom_file + msg_aria)
FileNotFoundError: [Errno 2] No such file or directory: '/media/...../dolphin/new/mintpy_test/inputs/geometryGeo.h5'"

**Whereas mintpy 1.6.3 fails at the prep_isce.py step does,** 

"load_data.py --template /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/mintpy_test/smallbaselineApp.cfg
processor : isce
SAR platform/sensor : unknown from project name "None"
--------------------------------------------------
prepare metadata files for isce products

**prep_isce.py -m /media/karthi/New_VolumeG/InSAR/puri_coreg_test/reference/IW1.xml -g /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference -b /media/karthi/New_VolumeG/InSAR/puri_coreg_test/baselines -f "/media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/unwrapped/*.unw.tif" --geom-files hgt.rdr.full lat.rdr.full lon.rdr.full los.rdr.full los.rdr.full shadowMask.rdr.full**
extract metadata from ISCE/topsStack xml file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/reference/IW1.xml

Extract metadata from ISCE/topsStack xml file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/reference/IW1.xml
extract metadata from geometry files: ['hgt.rdr', 'lat.rdr', 'lon.rdr', 'los.rdr']
writing  /media/karthi/New_VolumeG/InSAR/puri_coreg_test/reference/data.rsc
preparing RSC file for geometry files
write file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/hgt.rdr.full.rsc
write file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/lat.rdr.full.rsc
write file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/lon.rdr.full.rsc
write file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/los.rdr.full.rsc
write file: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/shadowMask.rdr.full.rsc
read perp baseline time-series from /media/karthi/New_VolumeG/InSAR/puri_coreg_test/baselines
**preparing RSC file for: /media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/unwrapped/*.unw.tif**
/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/osgeo/gdal.py:311: FutureWarning: Neither gdal.UseExceptions() nor gdal.DontUseExceptions() has been explicitly called. In GDAL 4.0, exceptions will be enabled by default.
  warnings.warn(
ERROR 4: **/media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/unwrapped/20240107_20240119.unw.tif' not recognized as being in a supported file format.**
/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/load_data.py:708: **UserWarning: prep_isce.py failed. Assuming its result exists and continue**
**warnings.warn('prep_isce.py failed. Assuming its result exists and continue...')**
searching geometry files info
input data files:
height          : /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/hgt.rdr.full
latitude        : /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/lat.rdr.full
longitude       : /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/lon.rdr.full
incidenceAngle  : /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/los.rdr.full
azimuthAngle    : /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/los.rdr.full
shadowMask      : /media/karthi/New_VolumeG/InSAR/puri_coreg_test/merged/geom_reference/shadowMask.rdr.full
**ERROR 4: `/media/karthi/New_VolumeG/InSAR/puri_coreg_test/dolphin/new/unwrapped/20240107_20240119.unw.tif' not recognized as being in a supported file format.**
Traceback (most recent call last):
  File "/home/karthi/miniconda3/envs/isce_env/bin/smallbaselineApp.py", line 10, in <module>
    sys.exit(main())
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/cli/smallbaselineApp.py", line 209, in main
    run_smallbaselineApp(inps)
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/smallbaselineApp.py", line 1155, in run_smallbaselineApp
    app.run(steps=inps.runSteps)
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/smallbaselineApp.py", line 908, in run
    self.run_load_data(sname)
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/smallbaselineApp.py", line 176, in run_load_data
    mintpy.cli.load_data.main(iargs)
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/cli/load_data.py", line 121, in main
    load_data(inps)
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/load_data.py", line 833, in load_data
    geom_geo_obj, geom_radar_obj = read_inps_dict2geometry_dict_object(iDict, geom_dset_name2template_key)
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/load_data.py", line 494, in read_inps_dict2geometry_dict_object
    atr = readfile.read_attribute(obsFiles[0])
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/utils/readfile.py", line 1320, in read_attribute
    atr.update(read_gdal_vrt(metafile))
  File "/home/karthi/miniconda3/envs/isce_env/lib/python3.10/site-packages/mintpy/utils/readfile.py", line 1706, in read_gdal_vrt
    atr['WIDTH']  = ds.RasterXSize
**AttributeError: 'NoneType' object has no attribute 'RasterXSize'**

**Reply by scottstanie** [reply]
thank you for the detail. i'm not sure what's going wrong, but MintPy is somehow assuming that you will have a geomtry HDF5 in geo coordiantes. I see you opened https://github.com/insarlab/MintPy/discussions/1460 , so hopefully someone more familiar with the how mintpy chooses loading functions/autodetects expected directory layouts can comment there

**Reply by Karthi-Wolf** [reply]
Okay Mr. @scottstanie  Thanks for your patience in this.
Surprisingly Mintpy's (1.6.3) **'info.py' considers dolphin's output in GEO coordinates**. The input SLCs (*.slc.full.vrt) to the dolphin were in radar coordinates only.

**info.py 20240212_20240319.unw.tif**
******************** Basic File Info ************************
file name: dolphin/unwrapped/20240212_20240319.unw.tif
file type: .tif
**coordinates : GEO**

**Comment by bcankara** [comment]
https://github.com/bcankara/Dolphin2MintPy try
